### PR TITLE
Consensus interface

### PR DESCRIPTION
**Pull request by olgavrou**
Related to #300 

- Deleted nullreplicator.h and moved it's methods into pbft.h
- made kv::Replicator the base class for Raft and Pbft
- cleaned up some of the replicators that were not being used
- in nodestate.h replaced raft and pbft with replicator and we either setup raft or pbft, not both
- on_request is now a method and not a callback

Notes:
- This is not done and the methods moved from NullReplicator into Pbft do not return useful information yet.
- Also not all methods defined in Replicator are used by both pbft and raft, but it is a good first step for converging the two interfaces and making it more clear as to what will need to be further refactored down the road.
- add_configuration is a bit weird right now as it needs to be called by both pbft and raft and they currently need different info to set up their nodes/replicas

### CODE REVIEW

**Comment by cimetrics** [comment]
Performance report: 
 images 
 **Have a look before merging** :bar_chart: :eyes: :thumbsup:

File: src/kv/kvtypes.h [hunk 1]
```diff
@@ -116,6 +101,50 @@ namespace kv
     virtual crypto::Sha256Hash get_root() = 0;
   };
 
+  class Replicator
```
Comment (jumaffre) [inline src/kv/kvtypes.h]: As discussed, the `kv::Replicator` should only include the methods that the kv store requires.

File: src/kv/replicator.h [hunk 2]
```diff
@@ -191,6 +76,35 @@ namespace kv
     {
       return true;
     }
+
+    bool on_request(kv::TxHistory::RequestCallbackArgs args) override
```
Comment (jumaffre) [inline src/kv/replicator.h]: As per my other comment, this should go as the `StubReplicator` should only derive from `kv::Replicator`.

File: src/node/nodestate.h [hunk 3]
```diff
@@ -311,8 +307,8 @@ namespace ccf
         return Fail<StartNetwork::Out>("Deserialisation of tx0 failed");
 
       // Become the leader and force replication.
-      raft->force_become_leader();
-      raft->replicate({{1, protected_tx0, true}});
+      replicator->force_become_leader();
```
Comment (jumaffre) [inline src/node/nodestate.h]: As discussed, `replicator` should become `consensus`.

File: src/pbft/pbft.h [hunk 4]
```diff
@@ -93,27 +93,21 @@ namespace pbft
     NodeId id;
   };
 
-  template <class ChannelProxy>
+  template <class LedgerProxy, class ChannelProxy>
   class Pbft : public kv::Replicator
```
Comment (jumaffre) [inline src/pbft/pbft.h]: `Pbft` should inherit from the new intermediate class that needs to be created.

File: src/pbft/pbft.h [hunk 5]
```diff
@@ -211,79 +207,84 @@ namespace pbft
...
+      LOG_DEBUG_FMT("PBFT sending request {}", args.rid);
+      return client_proxy->send_request(
+        args.rid,
+        request_buffer,
+        sizeof(request_buffer),
+        rep_cb,
+        client_proxy.get());
     }
 
     // TODO(#PBFT): PBFT consensus should implement the following functions to
```
Comment (jumaffre) [inline src/pbft/pbft.h]: Maybe worth referring to https://github.com/microsoft/CCF/issues/57 in the comment?

File: src/consensus/pbft/pbft.h [hunk 6]
```diff
@@ -148,7 +144,7 @@ namespace pbft
 
       // Adding myself
       PrincipalInfo my_info;
-      my_info.id = local_id;
```
Comment (achamayou) [inline src/consensus/pbft/pbft.h]: Not very fond of replacing local_id with self, local_id was clearly the id of the local node, whereas self could be just about anything.

File: src/kv/kvtypes.h (hunk 1 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Do we need to keep a separate replicator if we have a generic consensus interface?

File: src/kv/kvtypes.h [hunk 7]
```diff
@@ -116,6 +101,50 @@ namespace kv
...
+      const NodeConf& node_conf = {}) = 0;
+
+    virtual void force_become_leader() = 0;
+    virtual void force_become_leader(
+      Version index,
+      Term term,
+      const std::vector<Version>& terms,
+      Version commit_idx_) = 0;
+
+    virtual void enable_all_domains() = 0;
```
Comment (achamayou) [inline src/kv/kvtypes.h]: This is clearly much beyond the scope of a Replicator.

File: src/consensus/pbft/pbft.h [hunk 8]
```diff
@@ -93,27 +93,21 @@ namespace pbft
...
     enclave::RPCSessions& rpcsessions;
     std::unique_ptr<raft::LedgerEnclave> ledger;
-
-    struct NodeConfiguration
-    {
-      NodeId node_id;
-      std::string host_name;
-      std::string port;
-    };
+    bool _is_leader;
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: Why does this have an underscore prefix?

File: src/consensus/pbft/pbft.h [hunk 9]
```diff
@@ -93,27 +93,21 @@ namespace pbft
     NodeId id;
   };
 
-  template <class ChannelProxy>
+  template <class LedgerProxy, class ChannelProxy>
   class Pbft : public kv::Replicator
   {
   private:
-    NodeId local_id;
+    NodeId self;
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: I think `local_id` was a clearer name.

File: src/kv/kvtypes.h (hunk 7 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: moving everything except replicate() into a separate class the Raft and Pbft will inherit from as we speak

File: src/consensus/pbft/pbft.h (hunk 6 above)
Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: Thanks I'll change it back. I was undecided on whether to change it or not :)

File: src/consensus/pbft/pbft.h [hunk 10]
```diff
@@ -211,79 +207,84 @@ namespace pbft
...
-        };
-
-        LOG_DEBUG_FMT("PBFT sending request {}", args.rid);
-        return client_proxy->send_request(
-          args.rid,
-          request_buffer,
-          sizeof(request_buffer),
-          rep_cb,
-          client_proxy.get());
+    bool on_request(kv::TxHistory::RequestCallbackArgs args) override
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: Minor: Struct should be passed by `const&`, to avoid unnecessary copies.

File: src/kv/kvtypes.h [hunk 11]
```diff
@@ -5,13 +5,16 @@
...
+#include <chrono>
 #include <functional>
 #include <limits>
 #include <memory>
+#include <unordered_set>
 #include <vector>
 
 namespace kv
 {
+  using Index = int64_t;
```
Comment (eddyashton) [inline src/kv/kvtypes.h]: What's an Index? (How does it differ from a Version?)

Comment (eddyashton) [reply src/kv/kvtypes.h]: (Looks like its only used in Replicator, could it be `Replicator::ConfigurationIndex`?)

File: src/kv/replicator.h [hunk 12]
```diff
@@ -191,6 +76,35 @@ namespace kv
...
+    {
+      return true;
+    }
+    void periodic(std::chrono::milliseconds elapsed) override {}
+    bool is_follower() override
+    {
+      return false;
+    }
+    void recv_message(const uint8_t* data, size_t size) override {}
```
Comment (eddyashton) [inline src/kv/replicator.h]: Minor: Please add empty lines between function definitions.

File: src/kv/kvtypes.h (hunk 11 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Index is the Raft terminology (I almost wrote "term"!), as opposed to Version, which is the KV equivalent.

Comment (olgavrou) [reply src/kv/kvtypes.h]: Yes that was copied over from raft.h. I'll change it to ConfigIndex

Comment (achamayou) [reply src/kv/kvtypes.h]: Why config index? It should be the generic consensus term we adopt for both Raft and PBFT.

Comment (olgavrou) [reply src/kv/kvtypes.h]: add_configuration is a bit awkward right now since it is used both in pbft and in raft and they need different info to configure nodes. Index is only used by raft at the moment. pbft doesn't really have index in it's terminology at the moment.

File: src/kv/kvtypes.h (hunk 1 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: Renamed Replicator to Consensus and for now keeping all the related information there. See #304 for future refactoring/improvements

File: src/consensus/pbft/pbft.h [hunk 13]
```diff
@@ -50,8 +50,8 @@ namespace pbft
       NodeId to = principal.pid();
       if (to == id)
       {
-        // If a replica sends a message to itself (e.g. if f == 0), handle the
-        // message straight away without writing it to the ringbuffer
+        // If a replica sends a message to itlocal_id (e.g. if f == 0), handle
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: Find and replace error

File: src/consensus/pbft/pbft.h [hunk 14]
```diff
@@ -146,7 +142,7 @@ namespace pbft
       std::string pubk_enc =
         "893d4101c5b225c2bdc8633bb322c0ef9861e0c899014536e11196808ffc0d17";
 
-      // Adding myself
+      // Adding mylocal_id
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: Again.

**Comment by cimetrics** [comment]
Performance report: 
 images 
 **Have a look before merging** :bar_chart: :eyes: :thumbsup:

File: src/consensus/pbft/pbft.h (hunk 13 above)
Comment (achamayou) [inline src/consensus/pbft/pbft.h]: search & replace?

File: src/consensus/pbft/pbft.h [hunk 15]
```diff
@@ -174,7 +170,7 @@ namespace pbft
         0,
         pbft_network.get(),
         &message_receiver_base);
-      LOG_INFO_FMT("PBFT setup for self with id: {}", local_id);
+      LOG_INFO_FMT("PBFT setup for local_id with id: {}", local_id);
```
Comment (eddyashton) [inline src/consensus/pbft/pbft.h]: Minor: This message is weirdly worded now, perhaps just "PBFT setup for local_id: {}"?

File: src/consensus/pbft/pbft.h [hunk 16]
```diff
@@ -104,16 +104,10 @@ namespace pbft
...
     enclave::RPCSessions& rpcsessions;
     std::unique_ptr<raft::LedgerEnclave> ledger;
-
-    struct NodeConfiguration
-    {
-      NodeId node_id;
-      std::string host_name;
-      std::string port;
-    };
+    bool isleader;
```
Comment (achamayou) [inline src/consensus/pbft/pbft.h]: if we have local_id and my_info, we should have is_leader?

File: src/consensus/pbft/pbft.h (hunk 14 above)
Comment (achamayou) [inline src/consensus/pbft/pbft.h]: More search and replace?

File: src/consensus/pbft/pbft.h [hunk 17]
```diff
@@ -211,79 +207,84 @@ namespace pbft
...
-        LOG_DEBUG_FMT("PBFT sending request {}", args.rid);
-        return client_proxy->send_request(
-          args.rid,
-          request_buffer,
-          sizeof(request_buffer),
-          rep_cb,
-          client_proxy.get());
+    bool on_request(const kv::TxHistory::RequestCallbackArgs& args) override
+    {
+      LOG_INFO_FMT("in on request of pbft");
```
Comment (achamayou) [inline src/consensus/pbft/pbft.h]: in on_request?

Comment (achamayou) [reply src/consensus/pbft/pbft.h]: this debug, or more likely trace, not info.

File: src/kv/kvtypes.h [hunk 18]
```diff
@@ -5,13 +5,16 @@
...
+#include <chrono>
 #include <functional>
 #include <limits>
 #include <memory>
+#include <unordered_set>
 #include <vector>
 
 namespace kv
 {
+  using ConfigIndex = int64_t;
```
Comment (achamayou) [inline src/kv/kvtypes.h]: This is not a ConfigIndex, and Term also needs a rename.

File: src/consensus/pbft/pbft.h (hunk 17 above)
Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: thanks' forgot that there

File: src/kv/kvtypes.h [hunk 19]
```diff
@@ -116,6 +103,50 @@ namespace kv
...
+    virtual bool replicate(
+      const std::vector<std::tuple<Version, std::vector<uint8_t>, bool>>&
+        entries) = 0;
+    virtual Term get_term() = 0; // TODO(#api): this ought to have a more
+                                 // abstract name than Term
+
+    virtual Term get_term(Version version) = 0;
+    virtual Version get_commit_idx() = 0;
+
+    virtual NodeId leader() = 0;
```
Comment (eddyashton) [inline src/kv/kvtypes.h]: If we're making the concept of a Leader common to all Consensusesuseises, I think we should implement the shared logic here rather than duplicating/stubbing it elsewhere?

ie - given `leader()` and `id()`, I think we can implement `is_leader()` here as `return leader() == id()` (and similarly `is_follower()`, which should probably directly follow `is_leader()`)? And _maybe_ we could also make `leader` and `local_id` fields of this object and implement `leader()` and `id()` for everyone? And implement the 'obvious' bits of `force_become_leader`?

I think this is neater; the methods remain `virtual` for any implementor who disagrees/wants to extend, but the implementations we _currently_ have take less code.

**Comment by cimetrics** [comment]
Performance report: 
 images 
 **Have a look before merging** :bar_chart: :eyes: :thumbsup:

File: src/consensus/pbft/pbft.h (hunk 16 above)
Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: I agree but then we can't have a method named is_leader()

File: src/kv/kvtypes.h (hunk 19 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: Thanks, can you add this as a next step on the issue #300 so I can take it into consideration? The common interface will need more refactoring in general, trying to keep this PR as small as possible for now.

**Comment by cimetrics** [comment]
Performance report: 
 images 
 **Have a look before merging** :bar_chart: :eyes: :thumbsup:

File: src/kv/kvtypes.h (hunk 19 above)
Comment (eddyashton) [reply src/kv/kvtypes.h]: [Done](https://github.com/microsoft/CCF/issues/300#issuecomment-521213035).

File: src/kv/kvtypes.h (hunk 18 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: Agreed on SeqNo and View instead of Index (or ConfigIndex) and Term

File: src/consensus/pbft/pbft.h (hunk 16 above)
Comment (achamayou) [reply src/consensus/pbft/pbft.h]: We probably should not have that, and have a role() method instead, returning a Role enum with PRIMARY, BACKUP, CANDIDATE, UNKNOWN.

Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: In which case will it be CANDIDATE or UNKNOWN in raft or in pbft? And when will it transition to PRIMARY or BACKUP?

Comment (achamayou) [reply src/consensus/pbft/pbft.h]: See: https://github.com/microsoft/CCF/blob/6efad0b64cc4dd3d0b20fd93228472ad7e8b329d/src/raft/raft.h#L54 and its usage. Unknown may or may not be needed at startup.

File: src/kv/kvtypes.h (hunk 11 above)
Comment (achamayou) [reply src/kv/kvtypes.h]: Yep, let's align terminology as agreed.

File: src/consensus/pbft/pbft.h [hunk 20]
```diff
@@ -50,8 +50,8 @@ namespace pbft
       NodeId to = principal.pid();
       if (to == id)
       {
-        // If a replica sends a message to itself (e.g. if f == 0), handle the
-        // message straight away without writing it to the ringbuffer
+        // If a replica sends a message to local_id (e.g. if f == 0), handle
```
Comment (jumaffre) [inline src/consensus/pbft/pbft.h]: Minor: local_id -> itself?

File: src/consensus/pbft/pbft.h [hunk 21]
```diff
@@ -211,79 +207,84 @@ namespace pbft
...
+        sizeof(request_buffer),
+        rep_cb,
+        client_proxy.get());
     }
 
     // TODO(#PBFT): PBFT consensus should implement the following functions to
     // return meaningful information to clients (e.g. global commit, term/view)
-    // instead of relying on the NullReplicator
-    NodeId leader() override
+    // instead of relying on the NullConsensus, see
```
Comment (jumaffre) [inline src/consensus/pbft/pbft.h]: Is `NullConsensus` still a thing?

File: src/node/nodestate.h [hunk 22]
```diff
@@ -266,8 +258,8 @@ namespace ccf
       self = 0;
       setup_pbft();
       setup_history();
+      consensus->force_become_leader();
```
Comment (jumaffre) [inline src/node/nodestate.h]: Why is this necessary here?

File: src/node/rpc/frontend.h [hunk 23]
```diff
@@ -90,9 +90,9 @@ namespace ccf
 
     void update_raft()
```
Comment (jumaffre) [inline src/node/rpc/frontend.h]: As discussed, once we have implemented the [lightweight genesis model](https://github.com/microsoft/CCF/issues/178), we might not need to get the consensus from the store dynamically anymore but pass it to directly to the constructor of the frontend(s).

File: src/node/nodestate.h (hunk 22 above)
Comment (olgavrou) [reply src/node/nodestate.h]: Because frontend will check to see if we are the leader, see that we aren't and try to redirect the request to the leader, but there won't be one.

File: src/consensus/pbft/pbft.h (hunk 21 above)
Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: nope, removed the comment

**Comment by cimetrics** [comment]
Performance report: 
 images 
 **Have a look before merging** :bar_chart: :eyes: :thumbsup:

File: src/consensus/pbft/pbft.h (hunk 16 above)
Comment (olgavrou) [reply src/consensus/pbft/pbft.h]: Added as something to look into in next steps: #300

File: src/kv/kvtypes.h (hunk 18 above)
Comment (olgavrou) [reply src/kv/kvtypes.h]: Added as next step in #300
